### PR TITLE
fix(session-replay-browser): prevent 413 PAYLOAD_TOO_LARGE errors from losing event data

### PR DESCRIPTION
**Pull request by lewgordon-amplitude**
## Summary

Fixes [SR-3877](https://linear.app/amplitude/issue/SR-3877): 413 PAYLOAD_TOO_LARGE errors were silently dropping session replay data. Four root causes addressed:

- **Byte-accurate size measurement** (`base-events-store.ts`): replaced `str.length` (character count) with a zero-allocation UTF-8 byte counter. Non-ASCII content (Base64 images, emoji) was undercounted, letting batches exceed the 1 MB limit and guaranteeing 413s.
- **Reduced `MAX_EVENT_LIST_SIZE`** from 1,000,000 → 700,000 bytes to leave 30% headroom for JSON-serialization overhead (array brackets, commas, wrapping quotes, per-character escaping).
- **FullSnapshot size guard** (`event-compressor.ts`): FullSnapshots can reach 5–15 MB. If one exceeds `MAX_FULL_SNAPSHOT_SIZE` (20 MB), it is dropped with a warn log rather than sent and guaranteed to 413. The `onFullSnapshotProcessed` callback is still invoked so downstream flush logic runs correctly.
- **413 split-and-retry on both code paths**:
  - *Main thread* (`track-destination.ts`): `handlePayloadTooLargeResponse` splits the batch in half and re-queues both halves with fresh attempt counters. A `splitOnComplete` closure fires the original `onComplete` exactly once after both halves complete.
  - *Worker* (`worker/track-destination.ts`): `sendBatch` recursively splits on 413. Splits are sequential (not concurrent) to avoid amplifying load on a struggling endpoint. The 413 drop message is now prioritised over `MAX_RETRIES_EXCEEDED_MESSAGE` so the log accurately reflects the root cause.

## Test plan

- [ ] `packages/session-replay-browser` test suite passes: `cd packages/session-replay-browser && npx jest`
- [ ] UTF-8 byte-counting edge cases: surrogate pairs (4 bytes), 2-byte chars (é), 3-byte chars (€), orphaned surrogates
- [ ] FullSnapshot guard: oversized drop + warn, exact-boundary (20 MB allowed through), normal add
- [ ] Main-thread 413: single-event drop, multi-event split, `onComplete` fires exactly once
- [ ] Worker 413: split to halves, single-event drop, partial delivery, both-halves-fail, sequential depth
- [ ] Integration: single-event 413 → cannot-split-further warning; multi-event 413 → exactly 3 fetch calls

## Manual QA

- Covered entirely by automated tests. No UI component to verify manually.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes session replay batching/sending behavior (size caps, 413 handling, and FullSnapshot dropping), which can affect data delivery semantics and retry volume; mitigated by extensive new test coverage.
> 
> **Overview**
> Reduces 413 PAYLOAD_TOO_LARGE data loss by tightening payload sizing and making oversized sends recoverable.
> 
> Event batching now uses byte-accurate UTF-8 sizing (instead of `str.length`) and a smaller `MAX_EVENT_LIST_SIZE` (700 KB) to better account for JSON overhead. FullSnapshot events above `MAX_FULL_SNAPSHOT_SIZE` (20 MB) are dropped with a warning to avoid guaranteed 413s.
> 
> Both main-thread and worker send paths treat HTTP 413 as actionable: multi-event batches are split in half and retried (with `onComplete` fired once after both halves), while unsplittable single-event 413s are explicitly dropped with a clear warning. Tests add coverage for UTF-8 sizing edge cases, snapshot size guarding, and 413 split/retry behavior (including partial-delivery scenarios in the worker).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e942a1e0d8a8dfcfbcb9845051374a04990ad1be. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/events/base-events-store.ts [hunk 1]
```diff
@@ -55,10 +81,13 @@ export abstract class BaseEventsStore<KeyType> implements EventsStore<KeyType> {
...
-    // - Array brackets: [] = 2 characters
-    // - Commas between events: events.length - 1
-    // - Double quotes around each event: events.length * 2
+    // Approximate overhead from the array portion of the JSON payload:
+    // - Array brackets: [] = 2 bytes
+    // - Commas between events: events.length - 1 bytes
+    // - Double quotes wrapping each event string: events.length * 2 bytes
+    // Note: does not include the outer { version, events } wrapper (~22 bytes) or
+    // per-event JSON-escaping of " and \ characters; the reduced MAX_EVENT_LIST_SIZE
+    // cap (700 KB vs 1 MB) provides headroom for those.
```
Comment (stevenchien-amplitude) [inline packages/session-replay-browser/src/events/base-events-store.ts:90]: not blocking - more just curious. did 700kb end up being a sweet spot vs whatever other number we'd come up with for headway?

Comment (stevenchien-amplitude) [reply packages/session-replay-browser/src/events/base-events-store.ts:90]: oh nvm i saw the comment below. haha thanks!

**Review by stevenchien-amplitude — APPROVED** [review]
nice! lgtm
